File: python/remove_nth_node_from_end_of_list.py
# no.19
from utils import ListNode 
class Solution:
    def get_length(self, head):
        counter = 0
        curr = head
        while curr:
            counter += 1
            curr = curr.next
        return counter

    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        list_len = self.get_length(head)
        m = list_len - n
        curr = head
        if m == 0:
            return curr.next
        for i in range(m - 1):
            curr = curr.next
        curr.next = curr.next.next
        return head

File: C++/remove_nth_node_from_end_of_list.cpp
// no. 19
#include "linked_list.h"

class Solution {
public:
  int getLength(ListNode *head) {
    int counter = 0;
    ListNode *curr = head;
    while (curr) {
      counter++;
      curr = curr->next;
    }
    return counter;
  }
  ListNode *removeNthFromEnd(ListNode *head, int n) {
    int list_len = getLength(head);
    int m = list_len - n;
    ListNode *curr = head;
    if (m == 0) {
      return curr->next;
    }
    for (int i = 0; i < m - 1; i++) {
      curr = curr->next;
    }
    curr->next = curr->next->next;
    return head;
  }
};
